src/granite_orm/transactions.cr

```diff
@@ -34,100 +47,111 @@ module Granite::ORM::Transactions
       end
     end
 
+    private def __run_create
+      @created_at = @updated_at = Time.now.to_utc
+      fields = self.class.content_fields.dup
+      params = content_values
+      if value = @{{primary_name}}
+        fields << "{{primary_name}}"
+        params << value
+      end
+      begin
+        {% if primary_type.id == "Int32" %}
+          @{{primary_name}} = @@adapter.insert(@@table_name, fields, params, lastval: true).to_i32
+        {% elsif primary_type.id == "Int64" %}
+          @{{primary_name}} = @@adapter.insert(@@table_name, fields, params, lastval: true)
+        {% elsif primary_auto == true %}
+          {% raise "Failed to define #{@type.name}#save: Primary key must be Int(32|64), or set `auto: false` for natural keys.\n\n  primary #{primary_name} : #{primary_type}, auto: false\n" %}
+        {% else %}
+          if @{{primary_name}}
+            @@adapter.insert(@@table_name, fields, params, lastval: false)
+          else
+            message = "Primary key('{{primary_name}}') cannot be null"
+            errors << Granite::ORM::Error.new("{{primary_name}}", message)
+            raise DB::Error.new
+          end
+        {% end %}
```

Comment:
Could the macro code here for the primary key logic be moved into its own method?